pkg/proxy/middleware.go

```diff
@@ -0,0 +1,77 @@
+package proxy
+
+import (
+	"bytes"
+	"log"
+	"math/rand"
+	"net/http"
+	"time"
+)
+
+var _ http.Handler = &RetryMiddleware{}
+
+type RetryMiddleware struct {
+	nextHandler http.Handler
+	MaxRetries  int
+	rSource     *rand.Rand
+}
+
+func NewRetryMiddleware(maxRetries int, other http.Handler) *RetryMiddleware {
+	if maxRetries < 1 {
+		panic("invalid retries")
+	}
+	return &RetryMiddleware{
+		nextHandler: other,
+		MaxRetries:  maxRetries,
+		rSource:     rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (r RetryMiddleware) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	var capturedResp *responseBuffer
+	for i := 0; ; i++ {
+		capturedResp = &responseBuffer{
+			header: make(http.Header),
+			body:   bytes.NewBuffer([]byte{}),
```

Comment:
As we spoke about, I think it would be good to know whether we are adding to in-memory buffering or not by doing this. I think it is still worth having this feature even if we are making memory requirements worse. This makes me think that we might actually want to make retrying configurable:

* enabled/disabled
* list of status codes to retry? maybe this configurable item would be excessive?